db_master_deploy/deploy.sh

```diff
@@ -0,0 +1,331 @@
+    array=(${source_object//./ })
+    # check source objects
+    if [ "${#array[@]}" -ne "3" -a "${#array[@]}" -ne "1" ]; then
+        echo "table data sources have to be formatted like this: db.schema.table, database sources like this: db" >&2
+        exit 1
+    fi
+    # tables go here
+    if [ "${#array[@]}" -eq "3" ]; then
+        echo "processing table $source_object..."
+        source_db=${array[0]}
+        source_schema=${array[1]}
+        source_table=${array[2]}
+        target_db=${source_db%_*}"_"$target # lubis_3d_master -> lubis_3d_$target
+        target_schema="$source_schema"
+        target_table="$source_table"
+        source_id="$source_db.$source_schema.$source_table"
+        target_id="$target_db.$target_schema.$target_table"
+        array_target_table+=($target_id)
+        array_source_table+=($source_object)
+        array_target_combined+=($target_id)
+        check_table
+        check_source
+        copy_table
+    fi
+    # databases go here
+    if [ "${#array[@]}" -eq "1" ]; then
+        echo "processing database $source_object..."        
+        source_db=${array[0]}
+        target_db=${source_db%_*}"_"$target
+        target_db_tmp=$target_db""_tmp_deploy
+        array_target_db+=($target_db)
+        array_source_db+=($source_object)
+        array_target_combined+=($target_db)
+        check_database
+        check_source
+        bod_create_archive
+        copy_database        
+    fi    
+done
+
+END=$(date +%s%3N)
+echo "master has been updated in $((END-START)) milliseconds"
+
+# wait for all slaves
+while :
+do
+    diff=999
+    read slaves diff <<< $(psql \
+    -X \
+    -U pgkogis \
+    -h localhost \
+    -d postgres \
+    --single-transaction \
+    --set ON_ERROR_STOP=on \
+    --no-align \
+    -t \
+    --field-separator ' ' \
+    --quiet \
+    -c "select count(1) as slaves, coalesce(sum(pg_xlog_location_diff(sent_location,replay_location)),0) as diff FROM pg_stat_replication where state IN('streaming') ;")
+
+    # default value if connection above does not work
+    FOO=${diff:=999}
+
+    if [[ $diff -eq 0  && $slaves -eq $attached_slaves ]]
+    then
+        END_slaves=$(date +%s%3N)
+        echo "$slaves slaves have been updated in $((END_slaves-END)) milliseconds"
+        break
+    fi
+done
+echo "finished ${0##*/} $* in $((END_slaves-START)) milliseconds"
+
+# concatenate arrays for dml and ddl trigger
+source_db=$(IFS=, ; echo "${array_source_db[*]}")
+target_combined=$(IFS=, ; echo "${array_target_combined[*]}")
+
+# fire dml trigger in sub shell
+# redirect customized stdout and stderr to standard ones
+(
+bash "$MY_DIR/dml_trigger.sh" -s $target_combined -t $target 1>&5 2>&6
+)
+
+if [ "${#array_target_db[@]}" -gt "0" ]; then
+    # fire ddl trigger in sub shell
+    # redirect customized stdout and stderr to standard ones    
+    (
+    bash "$MY_DIR/ddl_trigger.sh" -s $source_db -t $target 1>&5 2>&6
+    )  
+fi
+
+rm -f "$lockfile"
```

Comment:
The lock file will be removed automatically if you CTRL-C the script or if there is any error between the two trap statements. The trap section is listening to SIG, INT and TERM Signals: http://www.davidpashley.com/articles/writing-robust-shell-scripts/

The postgres commands used for database and table copy will assure themselves with transactions that the database objects are not entering an unclear state. theoretically it is possible that a table copy, which will be split up to all available cores, fails somewhere. Each core opens a connection and therefore a transaction. If one of these transactions fails, we will end with a partially filled table. But you will be noticed by the script about that and can restart the script.

If there is an error in a list of database objects (table,databases), you will have to redeploy the whole list. But since a deploy is quite fast (some minutes) this can be easily done.
You can read the script output and check where the error occurred. The database objects are processed in the order of the input list. The re-deploy can continue where the first deploy failed, with a reduced list.
By the way we never had a retry feature in our current deploy scripts.